internal/provisioner/kops_provisioner_cluster.go

```diff
@@ -226,20 +192,13 @@ func (provisioner *KopsProvisioner) CreateCluster(cluster *model.Cluster, awsCli
 func (provisioner *KopsProvisioner) ProvisionCluster(cluster *model.Cluster, awsClient aws.AWS) error {
 	logger := provisioner.logger.WithField("cluster", cluster.ID)
 
-	kops, err := kops.New(provisioner.s3StateStore, logger)
-	if err != nil {
-		return errors.Wrap(err, "failed to create kops wrapper")
-	}
-	defer kops.Close()
+	logger.Info("Provisioning cluster")
 
-	err = kops.ExportKubecfg(cluster.ProvisionerMetadataKops.Name)
+	kopsClient, err := provisioner.getCachedKopsClient(cluster.ProvisionerMetadataKops.Name, logger)
 	if err != nil {
-		return errors.Wrap(err, "failed to export kubecfg")
+		return errors.Wrap(err, "failed to get kops client from cache")
 	}
-
-	kopsMetadata := cluster.ProvisionerMetadataKops
-
-	logger.Info("Provisioning cluster")
+	defer provisioner.invalidateCachedKopsClientOnError(err, cluster.ProvisionerMetadataKops.Name, logger)
```

Comment:
The intent here to only check `err` at the end for a non-nil value as provided by the `err` assignment on line 197, right? If for instance the call on line 211 produces an err, the cache may remain?

Reply:
This implementation will invalidate the kops cache if any error is encountered before this method returns. Although many of the errors have nothing to do with kops or the cache, I figured it was better to be safe than sorry.

In the future, it may make sense to introduce a second error value in these flows like `k8sErr` that we use any time something related to kops/k8s is done. This would require the code creator to be very careful to use the error correctly otherwise the cache may not be invalidated when it should be. Still, may be good to follow-up on if the caching works well.